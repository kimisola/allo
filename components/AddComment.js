import React from "react";
import { connect } from "react-redux";
import { sendComment } from "./ActionCreators";
import fire from "../src/fire";
import { db } from "../src/fire";
import Tick3 from "../images/tick3.png";
import Cancel from "../images/cancel.png";
import Plus from "../images/plusG.png";
import Upload from "../images/upload.png";

class AddComment extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            commentTags: { planning:false, process:false, risk:false, achived:false },
            isCommentWinShowed: false,
            newText: "",
            newImg: "",
            newTags: [],
            fileName: "",
        }
    }

    openCommentWindow = () => {
        this.resetToDefault();
        this.setState( prevState => {
            return { isCommentWinShowed: !prevState.isCommentWinShowed }
        });
    }

    resetToDefault = () => {
        this.setState({
            commentTags: { planning: false, process: false, risk: false, achived: false },
            newText: "",
            newImg: "",
            newTags: [],
            fileName: "",
        })
    }

    // resetToDefault = () => {
    //     this.setState( prevState => {
    //         let newtext = prevState.newText
    //         newtext = "";
    //         let newimg = prevState.newImg
    //         newimg = "";
    //         let tagsState = [ "planning", "process", "risk", "achived" ]
    //         let newtags = prevState.newTags;
    //         newtags = [];
    //         tagsState.forEach((item)=>{
    //             this.setState( prevState => {
    //                 let commentTagsCopy =  prevState.commentTags
    //                 if ( commentTagsCopy[item] ){
    //                     commentTagsCopy[item] = false
    //                 }   
    //                 return Object.assign({}, prevState, { 
    //                     commentTags: commentTagsCopy,
    //                     newText: newtext,
    //                     newImg: newimg,
    //                     newTags: newtags,
    //                 });
    //             });
    //         })
    //     });
    // }
    
    getTextValue = (event) => {
        const textValue = event.target.value
        this.setState({ newText: textValue });
    }

    selectTags = (selected) => {
        let tags = this.state.commentTags
        tags[selected] = !tags[selected]

        let tagsState = [ "planning", "process", "risk", "achived" ]
        let textTag = []
        tagsState.forEach((element) => {
            if ( tags[element] ) {
                textTag.push(element)
            }
        });
        this.setState({ newTags: textTag })
    }

    fileUpload = (event) => {
        const file = event.target.files[0]
        const storageRef = fire.storage().ref("image");
        const imgRef = storageRef.child(file.name)
        const fileTypes = ["image/jpeg", "image/png","image/gif"];
        let flag = false;

        this.setState(() => {
            let newfileName = file.name
            if (newfileName.length > 12) {
                let shortName = newfileName.slice(0, 8) + "..." + newfileName.slice(-4)
                return ({ fileName: shortName }) 
            } else {
                return ({ fileName: newfileName })
            }
        });

        imgRef.put(file)
        .then(() => {
            for (let i = 0; i < fileTypes.length; i++) {
                if ( file.type === fileTypes[i] ) {
                    flag = true
                    imgRef.getDownloadURL().then((url) => {
                        console.log(url)
                        this.setState({ newImg: url });
                    })
                }
            }
            if (!flag) {
                alert("Only support jpeg/png/gif type files.");
            }
        }).catch((error) => {
            console.error("Error removing document: ", error.message);
        })      
    }

    sendComment = () => {
        let index = this.props.index;
        let newText = this.state.newText;
        let newImg = this.state.newImg;
        let newTags = this.state.newTags;
        let fileName = this.state.fileName;
        let indexForItem = this.props.indexForItem;
        let docId = "";

        if ( newText === "" &&  fileName === "" && newTags.length === 0) {
            alert("You didn't enter anything!")
        } else {
            this.props.sendComment(index, newText, newImg, newTags)
            this.openCommentWindow();
         
            // const db = fire.firestore();
            let firebaseUid = "";
            if ( this.props.currentBoard !== "" ) {
                firebaseUid = this.props.currentBoard
            } else {
                firebaseUid = this.props.firebaseUid
            } 

            db.collection("Boards/" + firebaseUid + "/Lists").where("index", "==", ((index+1)*2)).get()
            .then( async (querySnapshot) => {
                docId =  querySnapshot.docs[0].id;

                const route = db.collection("Boards/" + firebaseUid + "/Lists/" + docId + "/Items").doc();
                db.collection("Boards/" + firebaseUid + "/Lists/" + docId + "/Items").orderBy("index").get()
                .then( async (querySnapshot2) => {
                    let doc2 = querySnapshot2.docs;
                    indexForItem = ((doc2.length+1)*2)
                    // this.props.dispatch({ type: "setIndexForItem", indexForItem})
                    if ( newImg === undefined ) {
                        newImg = "";
                    }
                    route.set({
                        img: newImg,
                        tags: newTags,
                        text: newText,
                        index: indexForItem,
                        owner: this.props.userDisplayName,
                        edited: this.props.userDisplayName,
                        ownerImg: this.props.userPhotoURL,
                        editorImg: this.props.userPhotoURL,
                    }).catch((error)=> {
                        console.log("Error writing document: ", error.message);
                    })
                }).catch((error) => {
                    console.log("Error writing document: ", error.message);
                })
            })
        }         
    }

    render(){
        return(
            <React.Fragment>
                <div className="addItem" style={{display: this.state.isCommentWinShowed ? 'block' : 'none' }}>               
                    <div className="tags">
                        <div className="tag planning" style={{backgroundColor: this.state.commentTags.planning ? "#ff9f1a" : "grey" }} onClick={ () => this.selectTags("planning") }>Planning</div>
                        <div className="tag process" style={{backgroundColor: this.state.commentTags.process ? "#00c2e0" : "grey" }} onClick={ () => this.selectTags("process") }>In Process</div>
                        <div className="tag risk" style={{backgroundColor: this.state.commentTags.risk ? "#eb5a46" : "grey" }} onClick={ () => this.selectTags("risk") }>At Risk</div>
                        <div className="tag achived" style={{backgroundColor: this.state.commentTags.achived ? "#4bbf6b" : "grey" }} onClick={ () => this.selectTags("achived") }>Achieved</div>
                    </div>
                    <div>
                        <textarea type="text" value={ this.state.newText } onChange={ this.getTextValue }></textarea>
                    </div>
                    <div className="addItemFooter">
                        <div className="addItemFooter2">
                            <div className="imgUploadDiv">
                                <label action="/somewhere/to/upload" encType="multipart/form-data">
                                    <div className="imgUpload">
                                        <img src={ Upload } />
                                    </div>
                                    <input name="progressbarTW_img" type="file" accept="image/gif, image/jpeg, image/png" onChange={ this.fileUpload } style={{display:'none' }} />    
                                </label>
                                <div className="fileNameDiv"> {this.state.fileName} </div>
                            </div>
                            <div className="addItemFeature">
                                <div className="addComment" onClick={ this.sendComment }> 
                                    <img src={ Tick3 } />
                                </div>
                                <div className="cancel">
                                    <img src={ Cancel } onClick={ this.openCommentWindow }/>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
                <div className="itemFooter">
                    <div className="add">
                        <img src={ Plus } onClick={ this.openCommentWindow }/>
                    </div>
                </div>
            </React.Fragment>
        )
    }
}

const mapStateToProps = (state , ownprops) => {
    return {
        index: ownprops.index,
        indexForItem: state.indexForItem,
        text: state.text,
        listTitle: state.listTitle,
        commentURL: state.commentURL,
        commentTags: state.commentTags,
        whichTheme: state.whichTheme,
        addNewCommentOpen: state.addNewCommentOpen,
        //commentWindow: state.commentWindow,
        userDisplayName: state.userDisplayName,
        userPhotoURL: state.userPhotoURL,
        firebaseUid: state.firebaseUid,
        currentBoard: state.currentBoard,
    }
}
const mapDispatchToProps = (dispatch) => {
    return {
        sendComment: (index, newText, newImg, newTags) => { dispatch(sendComment(index, newText, newImg, newTags)) },
    }
}
export default connect(mapStateToProps, mapDispatchToProps)(AddComment);