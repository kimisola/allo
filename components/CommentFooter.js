import React from "react";
import ReactDOM from "react-dom";
import { connect } from "react-redux";
import Plus from "../images/plus.png";


// 已廢

class ItemFooter extends React.Component {
    constructor(props){
        super(props);
    }


    creatComment = () => {
        console.log("run creat comment")
        console.log(i)

        this.props.dispatch({ type: "addComment" })

    }


    render(){
        return(
            <React.Fragment>
                
                <div className="itemFooter">
                    <div className="add">
                        <img src={ Plus } onClick={ () => creatComment()}/>
                    </div>
                </div>

            </React.Fragment>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        text: state.text,
        listTitle: state.listTitle,
        addNewCommentOpen:state.addNewCommentOpen
    }
}

export default connect(mapStateToProps)(ItemFooter);