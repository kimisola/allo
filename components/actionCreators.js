export function aCreatTheme() {
    return{
        type: "addList"
    }
}

export function aGetTitleValue(value) {
    return{
        type: "getNewTitleValue", 
        value
    }
}

export function aCreatTitle(newListTitle, newText) {
    return{
        type: "newRenderComments", 
        newListTitle,
        newText
    }
}

export function aSetUpComWin(myComWin) {
    return {
        type: "setUpComWin",
        myComWin
    }
}

export function aRenderComments(Data1, Data2) {
    return {
        type: "renderComments",
        Data1,
        Data2
    }
}

export function aSetCurrentUser(userDisplayName, userPhotoURL, userEmail, firebaseUid) {
    return {
        type: "setCurrentUser",
        userDisplayName, 
        userPhotoURL, 
        userEmail, 
        firebaseUid
    }
    
}