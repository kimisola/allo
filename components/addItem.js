import React from 'react';
import ReactDOM from 'react-dom';
import "../src/main.css";
import Tick from "../images/tick.png";
import Cross from "../images/cross.png";


class AddItem extends React.Component {
    constructor(props){
        super(props);
    }

    render(){
        return(
            <React.Fragment>
                
                    <div className="tags">
                        <div className="tag planning">Planning</div>
                        <div className="tag process">In Process</div>
                        <div className="tag risk">At Risk</div>
                        <div className="tag achived">Achieved</div>
                    </div>
                    <div>
                        <textarea type="text"></textarea>
                    </div>
                    <div className="addItemFooter">
                        <div className="imgUpload">
                            <form action="/somewhere/to/upload" encType="multipart/form-data">
                                <input name="progressbarTW_img" type="file" accept="image/gif, image/jpeg, image/png" />    
                            </form>
                        </div>
                        <div className="addItemFeature">
                            <div className="addComment">
                                <img src={ Tick } />
                            </div>
                            <div className="cancel">
                                <img src={ Cross } />
                            </div>
                        </div>
                    </div>
                

            </React.Fragment>
        )
    }
}
export default AddItem;