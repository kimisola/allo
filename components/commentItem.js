import React from 'react';
import ReactDOM from 'react-dom';
import Tag from "../images/tag.png";
import Pencil from "../images/pencil.png";
import GarbageCan from "../images/garbagecan.png";
import Cross from "../images/cross.png"
import Plus from "../images/plus.png";
import ListTitle from "../components/listTitle";
import AddItem from "../components/addItem";
import CommentMenu from "../components/commentMenu";
import { connect } from 'react-redux';
import fire from "../src/fire";



class CommentItem extends React.Component {
    constructor(props){
        super(props);
    }

    addComment = (i) => {
        this.props.dispatch({ type: "addComment", i })
    }

    render(){
        console.log("render list title", this.props.listTitle)
        console.log("render text", this.props.text)
        return(
            <React.Fragment>

            {this.props.listTitle.map((item , i) =>
                <React.Fragment>
                <div className="sectionWrapper" >
                    <div className="section">
                        
                        <ListTitle title={ item } indexWin={ i }/>
                           
                        <div className="comment">
                            {this.props.text[i].map((item, j) =>
                            
                                <div className="item" >
                                    <div className="itemHead">
                                        <div className="tags">

                                        {item.tags.map((tag) => {
                                            switch (tag) {
                                                case "planning":
                                                    return <div className="tag planning">Planning</div>

                                                case "process":
                                                    return <div className="tag process" >In Process</div>
                                            
                                                case "risk":
                                                    return <div className="tag risk">At Risk</div>
                                            
                                                case "achived":
                                                    return <div className="tag achived" >Achieved</div>

                                                    default:
                                                break;
                                                }
                                            })}
                                    
                                        </div>

                                        <CommentMenu  listId={`${i}`} comId={`${j}`}   />

                                    </div>                              
                                    
                                    <div className="itemBody">
                                        <div className="message">                       
                                            <div className="msgText"> {item.text} </div>         
                                            <div className="msgImg"> <img src={ item.img } /> </div>
                                        </div>
                                        {/* 可以獨立成一個 component 
                                        <div className="featureDiv">
                                            <div className="feature" style={{display: 'none' }}>
                                                <div className="edit">
                                                    <img src={ Pencil } />
                                                </div>
                                                <div className="trashcan">
                                                    <img src={ GarbageCan } />
                                                </div>
                                            </div>
                                        </div> */}
                                    </div>
                                </div>
                            )}
                        </div>
                        <div className="addItem" style={{display: this.props.commentWindow[i] ? 'block' : 'none' }}>
                        <AddItem />
                        </div>
                        <div className="itemFooter">
                            <div className="add">
                                <img src={ Plus } onClick={ () => this.addComment(i)}/>
                            </div>
                        </div>
                    </div>
                </div>

                </React.Fragment>  
            )}
          
            </React.Fragment>
        )
    }
}


const mapStateToProps = (state) => {
    return {
        text: state.text,
        listTitle: state.listTitle,
        addNewCommentOpen: state.addNewCommentOpen,
        deleteThemeConfirmOpen: state.deleteThemeConfirmOpen,
        whichWindowOpen: state.whichWindowOpen,
        commentWindow: state.commentWindow,
    }
}

export default connect(mapStateToProps)(CommentItem);