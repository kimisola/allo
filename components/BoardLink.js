import React from "react";
import { connect } from "react-redux";
import { BrowserRouter as Router, Switch, Route, Link } from "react-router-dom";
import { switchBoard } from"../actions/actionCreators"

class BoardLink extends React.Component {
    constructor(props){
        super(props);
    }

    switchBoard = () => {
        let targetLink = this.props.targetLink
        this.props.switchBoard(targetLink)
    }

    render(){
        let boardStyle = {
            board: {
                backgroundImage: `url(${this.props.boardBackground}) `,
                backgroundPosition: "center",
                backgroundSize: "cover",
                backgroundRepeat: "no-repeat",
                textDecoration: "none",
            },
            link: {
                color: "white", 
                textDecoration: "none",
            }
        };

        let targetURL = `/Board/${this.props.targetLink}`

        return(
            <React.Fragment>
                <div className="board" style={ boardStyle.board } onClick={ this.switchBoard } > 
                    <Link to={ targetURL } style={ boardStyle.link }>  { this.props.boardName } </Link>
                </div>
            </React.Fragment>
        )
    }
}

const mapStateToProps = (state, ownprops) => {
    return {
        firebaseUid: state.board.firebaseUid,
        userEmail: state.board.userEmail,
        userDisplayName: state.board.userDisplayName,
        userPhotoURL: state.board.userPhotoURL,
        boardBackground: ownprops.boardBackground,
        targetLink: ownprops.targetLink,
        boardName: ownprops.boardName,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        switchBoard: (targetLink) => { dispatch(switchBoard(targetLink)) },
    }
}
export default connect(mapStateToProps, mapDispatchToProps)(BoardLink);