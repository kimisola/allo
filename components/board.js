import React from "react";
import { connect, Provider } from "react-redux";
import fire from "../src/fire";
import { setUpComWin, renderComments, setIndexForTitle, loadingGifOn, switchBoard } from"../components/actionCreators"
import Topbar from "../components/topbar";
import SecondBar from "../components/secondBar";
import Section from "../components/section";
import loadingGif from "../images/loadingImg.gif";
import Gear from "../images/gear.png";


class Board extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            boardURL: "https://firebasestorage.googleapis.com/v0/b/allo-dc54c.appspot.com/o/homepageCover%2Fmaldives-1993704_1920.jpg?alt=media&token=b17d4f00-7e8f-4e2c-978f-c8ea14bb3a7f",
            exampleImg: "https://firebasestorage.googleapis.com/v0/b/allo-dc54c.appspot.com/o/image%2Fwelcome.png?alt=media&token=7a5e1d96-87c0-4a51-8dfc-e7c98b67579c",
            exampleAuthor: "https://firebasestorage.googleapis.com/v0/b/allo-dc54c.appspot.com/o/image%2F%E6%9C%AA%E5%91%BD%E5%90%8D.png?alt=media&token=264bf4c3-e1ed-42bd-8291-4928859932f7",
        }
    }

    componentDidMount() {
        
        console.log("run componentDidMount")
        let props = this.props;
        props.loadingGifOn();
        //read db
        const db = fire.firestore();
        
        let myDataTitle = [];
        let myDataText = [];
        let listsId = [];
        let Data = [];  // combine titles and texts
        let Data1 = [];  // store title
        let Data2 = [];  // store comment text

        let firebaseUid = "";
      
        if (firebaseUid == null) {  // 未登入
            window.location = "/";
        } 
        else {
            firebaseUid = this.props.match.params.id 
            console.log("this.props.match.params.id", this.props.match.params.id)
            this.props.switchBoard(this.props.match.params.id)
            getTitles(firebaseUid);
        }


        // set up or get background image
        db.collection("Boards").doc(firebaseUid).get()
        .then((querySnapshot) => {
            if( querySnapshot.data() !== undefined ){
                this.setState( prevState => {
                    let boardURL = prevState.boardURL
                    boardURL = querySnapshot.data().background
                    return { 
                        boardURL: boardURL,
                    }
                }); 
            } else {
                let ref = db.collection("Boards").doc(firebaseUid)
                ref.update({
                    background: "https://firebasestorage.googleapis.com/v0/b/allo-dc54c.appspot.com/o/homepageCover%2Fmaldives-1993704_1920.jpg?alt=media&token=b17d4f00-7e8f-4e2c-978f-c8ea14bb3a7f"
                }).then(() => {
                    console.log("Document successfully written!");
                }).catch((error) => {
                    console.error("Error removing document: ", error);
                })
            }
        })

        let userEmail = this.props.userEmail
        async function getTitles(firebaseUid) {  // 每次讀取資料庫就依照定義的 index 逐個抓出來再重新定義一次
            db.collection("Boards/").doc(firebaseUid).get().then((querySnapshot)=>{
                if (querySnapshot.data() === undefined ) {
                    let ref = db.collection("Boards").doc(firebaseUid)
                    ref.update({
                        owner : userEmail
                    }).then(() => {
                        console.log("Document successfully written!");
                    }).catch((error) => {
                        console.error("Error removing document: ", error.message);
                    })
                }
            })

            db.collection("Boards/" + firebaseUid + "/Lists").orderBy("index").get()
            .then(async (querySnapshot) => {
                let doc = querySnapshot.docs;

                for ( let i = 0; i < doc.length; i++ ) {       
                    listsId.push(doc[i].id)
                    let ref = db.collection("Boards/" + firebaseUid + "/Lists").doc(doc[i].id)
                    ref.update({
                        index: (((i+1)*2))  // 前後留空格讓之後移動可以有空間塞
                    })
                    myDataTitle.push(doc[i].data().title)
                    Data1.push(myDataTitle[i]);

                    // set an index value for next new added title
                    if ( i === doc.length - 1 ) {
                        let storeTitleIndex = ((doc.length+1)*2)
                        props.setIndexForTitle(storeTitleIndex)
                    }
                }
                getCommentText();
            });

            async function getCommentText(){
                for(let i = 0; i < listsId.length; i++ ) {  //讀取每一個 list 底下的 items
                    await db.collection("Boards/" + firebaseUid + "/Lists/" + listsId[i] + "/Items").orderBy("index").get()
                    .then((querySnapshot2) => {
                        let doc2 = querySnapshot2.docs;
                        for ( let j = 0; j < doc2.length; j++ ) {
                            let ref = db.collection("Boards/" + firebaseUid + "/Lists/" + listsId[i] + "/Items").doc(doc2[j].id)
                            ref.update({
                                index: (((j+1)*2)), // 前後留空格讓之後移動可以有空間塞
                                // edited: "",
                                // owner: ""
                            })           
                            myDataText.push(doc2[j].data())
                        }
                        Data2.push(myDataText);
                        myDataText = []; //reset comments under certain title
                    })
                } combineData();
            }
        }

        function combineData() { 
            for (let k = 0; k < Data1.length; k++) {
                Data.push(Data1[k]);
                Data.push(Data2[k]);
            }
            props.renderComments(Data1, Data2);
        };
    }

    componentDidUpdate(prevProps){
        let props = this.props;
        let currentBoard = this.props.currentBoard
        if ( currentBoard !==  prevProps.currentBoard) {
            props.loadingGifOn();
            //read db
            const db = fire.firestore();
            let firebaseUid = currentBoard;

            // set up or get background image
            db.collection("Boards").doc(firebaseUid).get()
            .then((querySnapshot) => {
                if( querySnapshot.data() !== undefined ){
                    this.setState( prevState => {
                        let boardURL = prevState.boardURL
                        boardURL = querySnapshot.data().background
                        return { 
                            boardURL: boardURL,
                        }
                    }); 
                } else {
                    let ref = db.collection("Boards").doc(firebaseUid)
                    ref.set({
                        background: "https://firebasestorage.googleapis.com/v0/b/allo-dc54c.appspot.com/o/homepageCover%2Fmaldives-1993704_1920.jpg?alt=media&token=b17d4f00-7e8f-4e2c-978f-c8ea14bb3a7f"
                    }).then(() => {
                        props.renderComments(["Welcome to a-llo guide !","List"], [
                            [{
                            edited:"a-llo",
                            editorImg: this.state.exampleAuthor,
                            img:"" ,
                            index:0,
                            owner:"a-llo",
                            ownerImg:this.state.exampleAuthor,
                            tags:[],
                            text:`To add your new list, press the
                            plus(+) button in the upper right corner of the page.
                            
                            At the bottom of the list, use the plus button to create cards which can include texts, images, and
                            tags.
                            
                            Need to edit or delete 
                            your cards? 
                            Click the upper right of the card.
                            
                            Small tips here, 
                            try dragging-dropping your lists or cards to manage schedule 
                            even more quickly.
                            
                            Now enjoy it !`
                            }],
                            [{
                            edited:"a-llo",
                            editorImg:this.state.exampleAuthor,
                            img: this.state.exampleImg,
                            index:2,
                            owner:"a-llo",
                            ownerImg:this.state.exampleAuthor,
                            tags:["planning"],
                            text:"",
                            },
                            {
                            edited:"a-llo",
                            editorImg:this.state.exampleAuthor,
                            img:"" ,
                            index:4,
                            owner:"a-llo",
                            ownerImg:this.state.exampleAuthor,
                            tags:["process","risk"],
                            text:`card.` 
                            },
                            {
                            edited:"a-llo",
                            editorImg:this.state.exampleAuthor,
                            img:"" ,
                            index:6,
                            owner:"a-llo",
                            ownerImg:this.state.exampleAuthor,
                            tags:["process","risk","achived"],
                            text:`card.`
                            }]
                        ]);
                        // 新增初始範例 ↓
                        let ref = db.collection("Boards/"+ firebaseUid + "/Lists").doc("alloExample")
                        ref.set({
                            title: "Welcome to a-llo guide !",
                            index: 2
                        }).then(() => {
                            console.log("Document successfully written!");
                            let ref = db.collection("Boards/"+ firebaseUid + "/Lists/alloExample/Items").doc()
                            ref.set({
                                edited:"a-llo",
                                editorImg:this.state.exampleAuthor,
                                img:"" ,
                                index:2,
                                owner:"a-llo",
                                ownerImg:this.state.exampleAuthor,
                                tags:[],
                                text:`To add your new list, press the
                                plus(+) button in the upper right corner of the page.
                                
                                At the bottom of the list, use the plus button to create cards which can include texts, images, and
                                tags.
                                
                                Need to edit or delete 
                                your cards? 
                                Click the upper right of the card.
                                
                                Small tips here, 
                                try dragging-dropping your lists or cards to manage schedule 
                                even more quickly.
                                
                                Now enjoy it !`
                            }).then(()=>{
                                console.log("Document successfully written!");
                            }).catch((error) => {
                                console.error("Error removing document: ", error);
                            })
                        }).catch((error) => {
                            console.error("Error removing document: ", error);
                        })                      
                        console.log("Document successfully written!");
                    }).then(()=>{
                        let ref = db.collection("Boards/"+ firebaseUid + "/Lists").doc("alloExample2")
                        ref.set({
                            title: "List",
                            index: 4
                        }).then(() => {
                            let ref = db.collection("Boards/"+ firebaseUid + "/Lists/alloExample2/Items").doc()
                            ref.set({
                                edited:"a-llo",
                                editorImg:this.state.exampleAuthor,
                                img:this.state.exampleImg,
                                index:2,
                                owner:"a-llo",
                                ownerImg:this.state.exampleAuthor,
                                tags:["planning"],
                                text:"",
                            })
                            let ref2 = db.collection("Boards/"+ firebaseUid + "/Lists/alloExample2/Items").doc()
                            ref2.set({
                                edited:"a-llo",
                                editorImg:this.state.exampleAuthor,
                                img:"" ,
                                index:4,
                                owner:"a-llo",
                                ownerImg:this.state.exampleAuthor,
                                tags:["process","risk"],
                                text:`card`
                            })
                            let ref3= db.collection("Boards/"+ firebaseUid + "/Lists/alloExample2/Items").doc()
                            ref3.set({
                                edited:"a-llo",
                                editorImg:this.state.exampleAuthor,
                                img:"" ,
                                index:6,
                                owner:"a-llo",
                                ownerImg:this.state.exampleAuthor,
                                tags:["process","risk","achived"],
                                text:`card`
                            })
                        })
                    })
                    .catch((error) => {
                        console.error("Error removing document: ", error);
                    })
                }
            })
           
            let myDataTitle = [];
            let myDataText = [];
            let listsId = [];
            let Data = [];  // combine titles and texts
            let Data1 = [];  // store title
            let Data2 = [];  // store comment text
 
            getTitles(firebaseUid);
            async function getTitles(firebaseUid) {  // 每次讀取資料庫就依照定義的 index 逐個抓出來再重新定義一次

                db.collection("Boards/" + firebaseUid + "/Lists").orderBy("index").get()
                .then(async (querySnapshot) => {
                    let doc = querySnapshot.docs;
    
                    for ( let i = 0; i < doc.length; i++ ) {       
                        listsId.push(doc[i].id)
                        let ref = db.collection("Boards/" + firebaseUid + "/Lists").doc(doc[i].id)
                        console.log("getTitles(firebaseUid);",firebaseUid);

                        ref.update({
                            index: (((i+1)*2)),  // 前後留空格讓之後移動可以有空間塞
                        })

                        myDataTitle.push(doc[i].data().title)
                        Data1.push(myDataTitle[i]);
    
                        // set an index value for next new added title
                        if ( i === doc.length - 1 ) {
                            let storeTitleIndex = ((doc.length+1)*2)
                            props.setIndexForTitle(storeTitleIndex)
                        }
                    }
                    getCommentText();
                });
    
                async function getCommentText(){
                    for(let i = 0; i < listsId.length; i++ ) {
                        await db.collection("Boards/" + firebaseUid + "/Lists/" + listsId[i] + "/Items").orderBy("index").get()
                        .then((querySnapshot2) => {
                            let doc2 = querySnapshot2.docs;
                            for ( let j = 0; j < doc2.length; j++ ) {
                                let ref = db.collection("Boards/" + firebaseUid + "/Lists/" + listsId[i] + "/Items").doc(doc2[j].id)
                                ref.update({
                                    index: (((j+1)*2)),  // 前後留空格讓之後移動可以有空間塞
                                })           
                                myDataText.push(doc2[j].data())
                            }
                            Data2.push(myDataText);
                            console.log(Data2,"doc2[j].data()doc2[j].data()doc2[j].data()")
                            myDataText = []; //reset comments under certain title
                        })
                    } combineData();
                }
            }
    
            function combineData() { 
                for (let k = 0; k < Data1.length; k++) {
                    Data.push(Data1[k]);
                    Data.push(Data2[k]);
                }
                props.renderComments(Data1, Data2);
            };
        }
    }

    fileUpload = (event) => {
        const file = event.target.files[0]
        console.log(event.target.files[0])
        var reader = new FileReader(); 
        const storageRef = fire.storage().ref("homepageCover");
        const imgRef = storageRef.child(file.name)
        const fileTypes = ["image/jpeg", "image/png","image/gif"]; 
        console.log("typetypetypetypetype",file.size)
        let flag = false;
        
            imgRef.put(file)
            .then((snapshot) => {
                for (let i = 0; i < fileTypes.length; i++) {
                    if ( file.type == fileTypes[i] ) { 
                        flag = true
                        if (file.size > 190000 ) {
                        console.log("Uploaded a blob or file!");
                        imgRef.getDownloadURL().then( async (url) => {
                            console.log(url)
                            this.setState( prevState => {
                                let boardURL = url
                                return Object.assign({}, prevState, {
                                    boardURL: boardURL,
                                })
                            });
                            const db = fire.firestore();
                            let firebaseUid = this.props.firebaseUid
                            db.collection("Boards").doc(firebaseUid)
                            .update({
                                background: this.state.boardURL
                            }).then(() => {
                                console.log("Document successfully written!")
                            }).catch((error)=> {
                                console.log("Error writing document: ", error);
                            })
                        })
                        } else { 
                            alert("Oops! Low resolution image.")
                            break;
                        }
                    }  
                }
                if (!flag) {
                    alert("Only support jpeg/png/gif type files.");
                }
            }).catch((error) => {
                console.error("Error removing document: ", error);
        })      
    }

    render(){

        const style = {
            backgroundStyle: {
                backgroundImage: `url("${this.state.boardURL}")`,
                backgroundPosition: 'center',
                backgroundSize: 'cover',
                backgroundRepeat: 'no-repeat'
            }
        };

        return(
            <React.Fragment>

                <main  style={ style.backgroundStyle} >
                    <div className="loading"  style={{display: this.props.isBoardLoaded ? 'none' : 'block' }} > 
                        <img src={ loadingGif } />
                    </div>
                    <div className="view" style={{display: this.props.isBoardLoaded ? 'block' : 'block' }} >
                        <Topbar />
                        <SecondBar />
                        <Section />
                    </div>
                    <label action="/somewhere/to/upload" encType="multipart/form-data" className="uploadBackground">
                        <img src={ Gear } />
                        <input name="progressbarTW_img" type="file" accept="image/gif, image/jpeg, image/png" onChange={ this.fileUpload } style={{display:'none' }} />
                    </label>
                </main>             
 
            </React.Fragment>
        )
    }
}

const mapStateToProps = (state) => {
    console.log("map state", state.firebaseUid);
    return {
        isBoardLoaded: state.isBoardLoaded,
        text: state.text,
        listTitle: state.listTitle,
        deleteThemeConfirmOpen: state.deleteThemeConfirmOpen,
        whichWindowOpen: state.whichWindowOpen,
        commentWindow: state.commentWindow,
        isLoggedIn: state.isLoggedIn,
        userEmail: state.userEmail,
        userDisplayName: state.userDisplayName,
        userPhotoURL: state.userPhotoURL,
        firebaseUid: state.firebaseUid,
        useruid: state.useruid,
        currentBoard: state.currentBoard
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        setUpComWin: (myComWin) => { dispatch(setUpComWin(myComWin)) },
        renderComments: (Data1, Data2) => { dispatch(renderComments(Data1, Data2)) },
        setIndexForTitle: (storeTitleIndex) => { dispatch(setIndexForTitle(storeTitleIndex))},
        loadingGifOn: () => { dispatch(loadingGifOn()) },
        switchBoard: (targetLink) => { dispatch(switchBoard(targetLink)) },
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(Board)