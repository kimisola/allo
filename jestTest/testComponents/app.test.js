import React from "react";
import Enzyme, {shallow} from "enzyme";
import Adapter from "enzyme-adapter-react-16";
Enzyme.configure({adapter: new Adapter()});
import {BoardLink} from "../../components/BoardLink";

const mockProps = {
  firebaseUid: "123456789",
  userEmail: "jestTest@test.com",
  userDisplayName: "Jest Test-1",
  userPhotoURL: "https://firebasestorage.googleapis.com/v0/b/allo-dc54c.appspot.com/o/image%2FalloImage.png?alt=media&token=59ffb5bb-4322-4a60-b331-d88e9aa4cbae",
  boardBackground: "https://images.unsplash.com/photo-1571649454759-c0de91fe6044?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1050&q=80",
  targetLink: "https://allo-dc54c.firebaseapp.com",
  boardName: "jest-Test",
}

describe('to match snapshot', () => {
  const component = shallow(<BoardLink {...mockProps} />);
  it("match snapshot", () => {
    expect(component).toMatchSnapshot();
  });

  it("must have boardName info", () => {
    expect(component.find(".board").children()).toHaveLength(1);
    expect(
      component
        .find(".board")
        .childAt(0)
        .text()
    ).toEqual("jest-Test");
  });
});