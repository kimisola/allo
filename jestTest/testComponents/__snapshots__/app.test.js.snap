// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`to match snapshot match snapshot 1`] = `ShallowWrapper {}`;
